src/TorchSharp/NN/Module.cs

```diff
@@ -325,55 +208,62 @@ protected internal virtual Module _to(ScalarType dtype)
-                                field.SetValue(this, p);
-                                ConditionallyRegisterParameter(fieldName, p);
-                                alreadyHandled.Add(p.handle);
-                                break;
-                            }
+                    var fieldsByComponentName = GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)
+                                                            .ToDictionary(field => field.ComponentName());
 
-                        case Tensor tensor when dtype == tensor.dtype:
-                            alreadyHandled.Add(tensor.handle);
-                            continue;
+                    foreach (var (name, param) in named_parameters(false).ToList()) {
+                        if (!param.toWillCopy(dtype ?? param.dtype, device ?? param.device)) continue;
 
-                        case Tensor tensor: {
-                                var t = tensor.to(dtype);
-                                field.SetValue(this, t);
-                                ConditionallyRegisterBuffer(fieldName, t);
-                                alreadyHandled.Add(t.handle);
-                                break;
-                            }
-                        }
-                    }
+                        // Store the requires_grad flag ahead, since we dispose the parameter after moving
+                        bool requiresGrad = param.requires_grad;
+                        Parameter p;
+                        // When moving the parameter, we don't want the autograd to track this movement on the graph.
+                        // In addition, we need the new tensor to be a leaf to accumulate gradients, so if we didn't
+                        // disable grad we would need to call .detach() on the moved tensor.
+                        using (var d = torch.no_grad())
+                            p = new Parameter(param.to(dtype ?? param.dtype, device ?? param.device, disposeAfter: true), requiresGrad);
+                        ConditionallyRegisterParameter(name, p);
 
-                    foreach (var (name, param) in named_parameters(false).ToList()) {
-                        if (alreadyHandled.Contains(param.handle)) continue;
-                        var t = param.to(dtype);
-                        ConditionallyRegisterParameter(name, t);
+                        // If this parameter is a field, set it
+                        if (fieldsByComponentName.TryGetValue(name, out var field))
+                            field.SetValue(this, p);
                     }
 
                     foreach (var (name, buffer) in named_buffers(false).ToList()) {
-                        if (alreadyHandled.Contains(buffer.handle)) continue;
-                        var t = buffer.to(dtype);
+                        if (!buffer.toWillCopy(dtype ?? buffer.dtype, device ?? buffer.device)) continue;
```

Comment:
Are old buffers/parameters disposed anywhere?

Reply:
Yup. In the ".to()" call I use the disposeAfter parameter.